src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,393 @@
+                    os.writeBytes(CipherCore.getKeyBytes(workItem));
+                }
+                return os.toByteArray();
+            }
+        } else if(keys != null) {
+                return null;
+        } else {
+            throw new InvalidKeyException(
+                "List of key segments could not be consolidated");
+        }
+    }
+
+    /**
+     * Perform the HKDF-Extract operation.
+     *
+     * @param inputKeyMaterial
+     *     the input keying material used for the HKDF-Extract operation.
+     * @param salt
+     *     the salt value used for HKDF-Extract; {@code null} if no salt value
+     *     is provided.
+     *
+     * @return a byte array containing the pseudorandom key (PRK)
+     *
+     * @throws InvalidKeyException
+     *     if an invalid salt was provided through the
+     *     {@code HkdfParameterSpec}
+     */
+    private byte[] hkdfExtract(byte[] inputKeyMaterial, byte[] salt)
+        throws InvalidKeyException, NoSuchAlgorithmException {
+
+        if (salt == null) {
+            salt = new byte[hmacLen];
+        }
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+        hmacObj.init(new SecretKeySpec(salt, hmacAlgName));
+
+        if (inputKeyMaterial == null) {
+            return hmacObj.doFinal();
+        } else {
+            return hmacObj.doFinal(inputKeyMaterial);
+        }
+    }
+
+    /**
+     * Perform the HKDF-Expand operation.
+     *
+     * @param prk
+     *     the pseudorandom key used for HKDF-Expand
+     * @param info
+     *     optional context and application specific information or {@code null}
+     *     if no info data is provided.
+     * @param outLen
+     *     the length in bytes of the required output
+     *
+     * @return a byte array containing the complete KDF output.  This will be at
+     *     least as long as the requested length in the {@code outLen}
+     *     parameter, but will be rounded up to the nearest multiple of the HMAC
+     *     output length.
+     *
+     * @throws InvalidKeyException
+     *     if an invalid PRK was provided through the {@code HKDFParameterSpec}
+     *     or derived during the extract phase.
+     */
+    private byte[] hkdfExpand(byte[] prk, byte[] info, int outLen)
+        throws InvalidKeyException, NoSuchAlgorithmException {
+        byte[] kdfOutput;
+        SecretKey pseudoRandomKey = new SecretKeySpec(prk, hmacAlgName);
+
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+
+        // Calculate the number of rounds of HMAC that are needed to
+        // meet the requested data.  Then set up the buffers we will need.
+        if (CipherCore.getKeyBytes(pseudoRandomKey).length < hmacLen) {
+            throw new InvalidKeyException(
+                "prk must be at least " + hmacLen + " bytes");
+        }
+        hmacObj.init(pseudoRandomKey);
+        if (info == null) {
+            info = new byte[0];
+        }
+        int rounds = (outLen + hmacLen - 1) / hmacLen;
```

Comment:
Perhaps make the addition as long, to rule out the risk of overflow?
```suggestion
        int rounds = (0L + outLen + hmacLen - 1) / hmacLen;
```